Tallies/TallyClerks/MGxsClerk_class.f90

```diff
@@ -0,0 +1,868 @@
+           N_2N, N_2Na, N_2Nd, N_2Nf, N_2Np, N_2N2a, N_2Nl(1):N_2Nl(16),  &
+           N_3N, N_3Na, N_3Nf, N_3Np, N_4N, N_Na, N_Np, N_Nd, N_Nt)
+
+        ! Find bin indexes
+        enIdx = self % energyMap % map(preColl)
+        if (allocated(self % spaceMap)) then
+          matIdx = self % spaceMap % map(preColl)
+        else
+          matIdx = 1
+        end if
+
+        ! Return if invalid bin index
+        if (enIdx == 0 .or. matIdx == 0) return
+
+        ! Calculate bin address
+        binIdx = self % energyN * (matIdx - 1) + enIdx
+        addr = self % getMemAddress() + self % width * (binIdx - 1) - 1
+
+        ! Score a scattering event from group g
+        call mem % score(preColl % wgt, addr + SCATT_EV_idx)
+
+        ! Get bin of outgoing energy
+        binEnOut = self % energyMap % map(postColl)
+
+        ! Return if invalid bin index
+        if (binEnOut == 0) return
+
+        ! Score scattering event from group g to g'
+        call mem % score(preColl % wgt, addr + SCATT_EV_idx + binEnOut)
+
+        ! Score outgoing scattering angle for P1 matrix
+        mu = muL * preColl % wgt
+        call mem % score(mu, addr + SCATT_EV_idx + self % energyN + binEnOut)
+
+        ! Score multiplicity matrix
+        prod = score * preColl % wgt
+        call mem % score(prod, addr + SCATT_EV_idx + 2*self % energyN + binEnOut)
+
+        ! Higher order scattering matrices
+        if (self % PN) then
+
+          ! Score outgoing scattering angle for P2 matrix
+          mu = HALF * (3.0_defReal*muL*muL - ONE) * preColl % wgt
+          call mem % score(mu, addr + SCATT_EV_idx + 3*self % energyN + binEnOut)
+
+          ! Score outgoing scattering angle for P3 matrix
+          mu = HALF * (5.0_defReal*muL**3 - 3.0_defReal*muL) * preColl % wgt
+          call mem % score(mu, addr + SCATT_EV_idx + 4*self % energyN + binEnOut)
+
+          ! Score outgoing scattering angle for P4 matrix
+          mu = (35.0_defReal*muL**4 - 30.0_defReal*muL*muL + 3.0_defReal)/8.0_defReal * preColl % wgt
+          call mem % score(mu, addr + SCATT_EV_idx + 5*self % energyN + binEnOut)
+
+          ! Score outgoing scattering angle for P5 matrix
+          mu = (63.0_defReal*muL**5 - 70.0_defReal*muL**3 + 15.0_defReal*muL)/8.0_defReal * preColl % wgt
+          call mem % score(mu, addr + SCATT_EV_idx + 6*self % energyN + binEnOut)
+
+          ! Score outgoing scattering angle for P6 matrix
+          mu = (231.0_defReal*muL**6 - 315.0_defReal*muL**4 + 105.0_defReal*muL*muL &
+                - 5.0_defReal)/16.0_defReal * preColl % wgt
+          call mem % score(mu, addr + SCATT_EV_idx + 7*self % energyN + binEnOut)
+
+          ! Score outgoing scattering angle for P7 matrix
+          mu = (429.0_defReal*muL**7 - 693.0_defReal*muL**5 + 315.0_defReal*muL**3 &
+                - 35.0_defReal*muL)/16.0_defReal * preColl % wgt
+          call mem % score(mu, addr + SCATT_EV_idx + 8*self % energyN + binEnOut)
```

Comment:
I may have spent too much time in deterministic land where this actually matters, but it might be a bit faster to precompute the powers up to 7 (or maybe just 3 or 4). It would save quite a lot of repeated multiplication/exponentiation.